File: src/minimum-window-substring/minimum-window-substring.py
# https://leetcode.com/problems/minimum-window-substring

'''
Key Insights:
    1) Did pretty good on this. Felt like a medium+ instead of a hard.
      Or maybe the practice is just working.

    2) On first attempt, I messed up setting the
      initial value of `size` to low. I set it to 1000, but input size
      could be 10^5. Basically did all the tricky programming correct, and
      goofed on an assumption.

    3) This is basically a harder version of
      'longest-substring-without-repeating-characters', but the premise of
      the solition is the same. Instead of combinining the properties of a
      set and a deque, you combine the properties of a dict and a dequeu
      (although I used Counter instead of dict for slight conveniance).
'''

import json
from operator import itemgetter
from collections import deque
from collections import Counter

class Solution(object):

    def valid_win(self, ds, dt):
        for key in dt:
            if key not in ds \
            or ds[key] < dt[key]:
                return False
        return True

    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        dt = Counter(t)
        ds = Counter()
        q = deque([])
        size = pow(10, 5) + 1
        res = ""
        for char in s:
            if char not in dt and len(ds) == 0:
                continue
            q.append(char)
            if char in dt:
                ds[char] += 1
                while ds[char] >= dt[char]:
                    if q[0] in dt and ds[q[0]] <= dt[q[0]]:
                        break
                    out = q.popleft()
                    ds[out] -= 1
            if self.valid_win(ds, dt):
                if len(q) < size:
                    size = len(q)
                    res = "".join([c for c in q])
        return res

if __name__ == "__main__":
    sol = Solution()
    tests = [
        ("ADOBECODEBANC", "ABC", "BANC"),
        ("a", "a", "a"),
        ("a", "aa", ""),
        itemgetter('s', 't', 'solution')(json.load(open("test1.json")))
    ]
    for (s, t, solution) in tests:
        result = sol.minWindow(s, t)
        assert result == solution, \
            f"result '{result}' != solution '{solution}'"
    print("✅ All tests passed")

File: src/minimum-window-substring/minimum-window-substring.rs
// https://leetcode.com/problems/minimum-window-substring
struct Solution;

use std::collections::{HashMap, VecDeque};

impl Solution {
    pub fn charfreq(word: &String) -> HashMap<char, usize> {
        let mut freq = HashMap::new();
        for ch in word.chars() {
            *freq.entry(ch).or_insert(0) += 1;
        }

        freq
    }

    pub fn valid_win(freq_win: &HashMap<char, usize>, freq_t: &HashMap<char, usize>) -> bool {
        for (ch, t_count) in freq_t.iter() {
            match freq_win.get(&ch) {
                None => return false,
                Some(win_count) => {
                    if win_count < t_count {
                        return false;
                    }
                }
            }
        }

        true
    }

    pub fn min_window(s: String, t: String) -> String {
        let freq_t = Self::charfreq(&t);
        let mut freq_win: HashMap<char, usize> = HashMap::new();
        let mut win: VecDeque<char> = VecDeque::new();
        let mut min_win: Option<String> = None;

        for ch in s.chars() {
            // println!("ch={:?}, freq_t={:?}, win={:?}, freq_win={:?}", ch, freq_t, win, freq_win);
            if !freq_t.contains_key(&ch) && win.is_empty() {
                continue;
            }
            win.push_back(ch);
            if freq_t.contains_key(&ch) {
                *freq_win.entry(ch).or_insert(0) += 1;
                while freq_win.get(&ch).unwrap_or(&0) >= freq_t.get(&ch).unwrap_or(&0) {
                    if freq_t.contains_key(&win[0]) && freq_win[&win[0]] <= freq_t[&win[0]] {
                        break;
                    }
                    let out = win.pop_front().unwrap();
                    if let Some(count) = freq_win.get_mut(&out) {
                        *count -= 1;
                        if *count == 0 {
                            freq_win.remove(&out);
                        }
                    }
                }
            }

            if Self::valid_win(&freq_win, &freq_t) {
                match min_win {
                    None => min_win = Some(win.iter().collect()),
                    Some(ref s) => {
                        if win.len() < s.len() {
                            min_win = Some(win.iter().collect())
                        }
                    }
                }
            }
        }

        min_win.unwrap_or("".to_string())
    }
}

fn main() {
    let tests = vec![
        (
            "ADOBECODEBANC".to_string(),
            "ABC".to_string(),
            "BANC".to_string(),
        ),
        ("a".to_string(), "a".to_string(), "a".to_string()),
        ("a".to_string(), "aa".to_string(), "".to_string()),
    ];

    for (s, t, solution) in tests {
        let result = Solution::min_window(s, t);
        assert_eq!(result, solution);
    }

    println!("✅ All tests passed")
}
